test/gui/blueprint.lua

```diff
@@ -0,0 +1,105 @@
+config = {
+    mode = 'fortress',
+}
+
+local b = reqscript('gui/blueprint')
+local gui = require('gui')
+local guidm = require('gui.dwarfmode')
+
+function test.fail_if_no_map_loaded()
+    local mock_is_map_loaded = mock.func(false)
+    mock.patch(dfhack, 'isMapLoaded', mock_is_map_loaded,
+        function()
+            expect.error_match('load a fortress map',
+                               function() b.BlueprintUI{}:show() end)
+        end)
+end
+
+local function send_keys(...)
+    local keys = {...}
+    for _,key in ipairs(keys) do
+        gui.simulateInput(dfhack.gui.getCurViewscreen(true), key)
+    end
+    delay()
+end
+
+local function load_ui()
+    -- remove screen mode switching once gui/blueprint supports it natively
+    for i=1,10 do
+        if df.global.ui.main.mode == df.ui_sidebar_mode.Default and
+                'dwarfmode/Default' == dfhack.gui.getCurFocus(true) then
+            send_keys('D_LOOK')
+            b.BlueprintUI{}:show()
+            delay()
+            return
+        end
+        send_keys('LEAVESCREEN')
+    end
+    error('Unable to get into look mode from current UI viewscreen.')
+end
+
+function test.minimal_happy_path()
+    local mock_print, mock_run_command = mock.func(), mock.func()
+    mock.patch({
+            {b, 'print', mock_print},
+            {dfhack, 'run_command', mock_run_command},
+            {gui, 'blink_visible', mock.func(true)},
+        },
+        function()
+            load_ui()
+            expect.eq('dfhack/lua', dfhack.gui.getCurFocus(true))
+            guidm.setCursorPos({x=10, y=20, z=30})
+            send_keys('SELECT')
+            send_keys('CURSOR_RIGHT', 'CURSOR_DOWN', 'CURSOR_DOWN',
+                      'CURSOR_UP_Z', 'CURSOR_UP_Z', 'CURSOR_UP_Z')
+            send_keys('SELECT')
+            expect.ne('dfhack/lua', dfhack.gui.getCurFocus(true))
```

Comment:
Might want to check for absence of a `dfhack/lua` or `dfhack/` prefix here (I don't think `dfhack_viewscreen::is_instance()` is exposed to Lua...), because Lua screens can add things to the focus string if they want. Not really sure if that's a valid concern here, though.

Glad the (minimal) mocking support is useful!

Reply:
done. and the mocking support has been great! it really simplifies the structure of these tests.

Reply:
Looks to me like you changed the other two calls to `expect.ne` but not this one.

Reply:
done.